### DISCUSSION TITLE
502 error and exec:sidekiq not found

### DISCUSSION THREAD

**Discussion by N-Yost** [mastodon/mastodon#23271] [Q&A] [answered]
Trying to start my own server on Ubuntu 22.10 with Linode, I am running into a 502 error. I followed the setup [here](https://docs.joinmastodon.org/admin/install/) and get everything installed relatively easy, but trying to go to my server returns a 502. 

I kept all of the defaults, but ran the  `$EDITOR /etc/systemd/system/mastodon-*.service` command anyways. At first, I got this response:
_-bash: /etc/systemd/system/mastodon-sidekiq.service: Permission denied_

After, giving execution permission (chmod 007 mastodon-sidekiq.service) I am now getting this error:
`/etc/systemd/system/mastodon-sidekiq.service: line 1: [Unit]: command not found
/etc/systemd/system/mastodon-sidekiq.service: line 5: [Service]: command not found
/etc/systemd/system/mastodon-sidekiq.service: line 13: exec: sidekiq: not found`

`journalctl -u mastodon-web` shows no entries
nginx error log:
`2023/01/27 02:02:37 [error] 32738#32738: *661 connect() failed (111: Connection refused) while connecting to upstream, client: <my ip>, server: <domain name>, request: "POST / HTTP/2.0", upstream: "http://127.0.0.1:3000/500.html", host: "<server ip>"`

If I had to guess, I would say it is a permission issue, but I'm not sure where. I would appreciate any help!

Edit:
Out of curiosity I tried finishing the walk through so I ran these two commands:
`systemctl daemon-reload
systemctl enable --now mastodon-web mastodon-sidekiq mastodon-streaming`

Once it was enabled, it would be a blank white page with a 500 error and the `journalctl -u mastodon-web` command had this:

Jan 27 02:12:00 fandomonium bundle[34079]: [94cdd9e0-ef76-436f-b2ed-88b18c5d7c5> method=GET path=/ format=html controller=HomeController action=index status=500 error='ActionView::Template::Error: Webpacker can't find media/images/preview.png in /home/mastodon/live/public/packs/manifest.json. Possible causes:
Jan 27 02:12:00 fandomonium bundle[34079]: 1. You want to set webpacker.yml valalue of compile to true for your environment
Jan 27 02:12:00 fandomonium bundle[34079]:    unless you are using the `webpack -w` or the webpack-dev-server.
Jan 27 02:12:00 fandomonium bundle[34079]: 2. webpack has not yet re-run to reflect updates.
Jan 27 02:12:00 fandomonium bundle[34079]: 3. You have misconfigured Webpacker's config/webpacker.yml file.
Jan 27 02:12:00 fandomonium bundle[34079]: 4. Your webpack configuration is not creating a manifest.
Jan 27 02:12:00 fandomonium bundle[34079]: Your manifest contains:
Jan 27 02:12:00 fandomonium bundle[34079]: {
Jan 27 02:12:00 fandomonium bundle[34079]: }
Jan 27 02:12:00 fandomonium bundle[34079]: ' duration=81.66 view=0.00 db=8.07

Checking the public directory, there was no packs sub-directory.

**Comment by ClearlyClaire** [comment, marked as answer]
Hi!

The `$EDITOR /etc/system/mastodon-*.service` is meant for you to edit the unit files in your default/favorite editor. It appears that `$EDITOR` was not set, which caused the shell to try and execute the unit files. This is not meant to happen and those files are not meant to be executed. I'd revert the execute bit if I were you.

The second issue seems to be caused by a failure to build the assets.

```shell
export RAILS_ENV=production
bundle exec rails assets:clobber && bundle exec rails assets:precompile
```

In `/home/mastodon/live` as the `mastodon` user. Report back if this errors out. Otherwise, restarting the services (`systemctl restart mastodon-{web,sidekiq}`) should do the trick.

**Reply by N-Yost** [reply]
Thank you for the quick reply! When I try to run the command, I'm getting this error and output:
```
Compiling...
Compilation failed:
node:internal/crypto/hash:71
  this[kHandle] = new _Hash(algorithm, xofLen);
                  ^

Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:71:19)
    at Object.createHash (node:crypto:133:10)
    at module.exports (/home/mastodon/live/node_modules/webpack/lib/util/createHash.js:135:53)
    at ConcatenatedModule._createIdentifier (/home/mastodon/live/node_modules/webpack/lib/optimize/ConcatenatedModule.js:563:16)
    at new ConcatenatedModule (/home/mastodon/live/node_modules/webpack/lib/optimize/ConcatenatedModule.js:445:27)
    at /home/mastodon/live/node_modules/webpack/lib/optimize/ModuleConcatenationPlugin.js:250:26
    at SyncBailHook.eval [as call] (eval at create (/home/mastodon/live/node_modules/tapable/lib/HookCodeFactory.js:19:10), <anonymous>:7:16)
    at SyncBailHook.lazyCompileHook (/home/mastodon/live/node_modules/tapable/lib/Hook.js:154:20)
    at /home/mastodon/live/node_modules/webpack/lib/Compilation.js:1351:37
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/home/mastodon/live/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:6:1)
    at AsyncSeriesHook.lazyCompileHook (/home/mastodon/live/node_modules/tapable/lib/Hook.js:154:20)
    at Compilation.seal (/home/mastodon/live/node_modules/webpack/lib/Compilation.js:1342:27)
    at /home/mastodon/live/node_modules/webpack/lib/Compiler.js:675:18
    at /home/mastodon/live/node_modules/webpack/lib/Compilation.js:1261:4
    at AsyncSeriesHook.eval [as callAsync] (eval at create (/home/mastodon/live/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:24:1)
    at AsyncSeriesHook.lazyCompileHook (/home/mastodon/live/node_modules/tapable/lib/Hook.js:154:20)
    at Compilation.finish (/home/mastodon/live/node_modules/webpack/lib/Compilation.js:1253:28)
    at /home/mastodon/live/node_modules/webpack/lib/Compiler.js:672:17
    at eval (eval at create (/home/mastodon/live/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:11:1)
    at /home/mastodon/live/node_modules/webpack/lib/Compilation.js:1185:12
    at /home/mastodon/live/node_modules/webpack/lib/Compilation.js:1097:9
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11) {
  opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
  library: 'digital envelope routines',
  reason: 'unsupported',
  code: 'ERR_OSSL_EVP_UNSUPPORTED'
}

Node.js v18.7.0
```

**Reply by ClearlyClaire** [reply]
Try with `export NODE_OPTIONS=--openssl-legacy-provider`

**Reply by N-Yost** [reply]
Hm, it is still failing, but isn't giving a reason for the failure now. 

```
I, [2023-01-27T17:37:02.933578 #82925]  INFO -- : Writing /home/mastodon/live/public/assets/pghero/application-c31338f656687c1d733bb0f48d40acd076e24060f3dcff83b34870e4ccc2789d.css
I, [2023-01-27T17:37:02.933725 #82925]  INFO -- : Writing /home/mastodon/live/public/assets/pghero/application-c31338f656687c1d733bb0f48d40acd076e24060f3dcff83b34870e4ccc2789d.css.gz
Compiling...
Compilation failed:

mastodon@<host name>:~/live$
```

**Reply by ClearlyClaire** [reply]
It may be running out of memory, try adding swap, or editing the configuration as follow:
```patch
diff --git a/config/webpack/production.js b/config/webpack/production.js
index 143a23b99..f53a1b3a4 100644
--- a/config/webpack/production.js
+++ b/config/webpack/production.js
@@ -13,6 +13,7 @@ const sharedConfig = require('./shared');
 const root = resolve(__dirname, '..', '..');
 
 module.exports = merge(sharedConfig, {
+  parallelism: 1,
   mode: 'production',
   devtool: 'source-map',
   stats: 'normal',
```

You'll need to do `bundle exec rails assets:clobber` before retrying.

**Reply by N-Yost** [reply]
That was the issue and it is now up and running. Thank you!!
